nck/readers/mytarget_reader.py

```diff
@@ -0,0 +1,252 @@
+        retries={
+            'max_attempts': 10
+        }
+    )
+)
+
+
+class MyTargetReader(Reader):
+
+    def __init__(
+        self,
+        client_id,
+        client_secret,
+        mail,
+        agency,
+        **kwargs
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.mail = mail
+        self.agency = agency
+        self.access_token = None
+        self.agency_client_token = None
+
+    def create_agency_token_request(self, agency_client):
+        url = 'https://target.my.com/api/v2/oauth2/token.json'
+        headers = {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'Host': 'target.my.com'
+        }
+        payload = dict(
+            grant_type='agency_client_credentials',
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            agency_client_name=agency_client
+        )
+        return dict(
+            url=url,
+            headers=headers,
+            data=payload
+        )
+
+    def retrieve_agency_client_response(self):
+        return requests.post(**self.create_agency_token_request(self.agency)).json()
+
+    def is_agency_client_token_new(self, rsp):
+        return "error" not in rsp.keys()
+
+    def __write_agency_client_token_s3(self, content: Dict[str, str]):
+        s3_resource.Object('token-mytarget', f'{self.agency_client_token}_last_token.json').put(Body=json.dumps(content))
+        logging.info('The token has been uploaded to the bucket.')
+
+    def __retrieve_agency_client_token_file(self) -> Dict[str, str]:
+        obj = s3_resource.Object('token-mytarget', f'{self.agency_client_token}_last_token.json')
+        logging.info('The last token has been retrieved from the bucket.')
+        return json.loads(obj.get()['Body'].read())
+
+    def retrieve_agency_client_token(self):
+        rsp = self.retrieve_agency_client_response()
+        if self.is_agency_client_token_new(rsp):
+            logging.info("New token retrieved, going to store it to s3")
+            self.__write_agency_client_token_s3(rsp)
+            return rsp
+        else:
+            logging.info("No more token available, going to retrieve the last one from s3")
+            return self.__retrieve_agency_client_token_file()
+
+    def set_agency_client_token(self):
+        self.agency_client_token = self.retrieve_agency_client_token()
+
+    def create_campaign_id_request(self, offset: int):
+        url = "https://target.my.com/api/v2/banners.json"
+        headers = {
+            'Authorization': 'Bearer ' + self.agency_client_token['access_token'],
+            'Host': 'target.my.com'
+        }
+        params = dict(offset=offset)
+        return dict(
+            url=url,
+            headers=headers,
+            params=params
+        )
+
+    def get_campaign_id_response(self, offset: int):
+        return requests.get(**self.create_campaign_id_request(offset)).json()
```

Comment:
Chaining methods like that is powerful but it prevents us from getting the response object that contains interesting information such as HTTP codes. They would be very helpful in case of errors for ex and we should be sure we log errors properly.